### ISSUE TITLE
Add example for off-pac

### ISSUE THREAD

**Issue by dantp-ai** [dantp-ai/gym-puddle#35]
## Status

Implemented on branch [`feature/off-pac-discrete-sac`](https://github.com/dantp-ai/gym-puddle/tree/feature/off-pac-discrete-sac) — commits [`6ac75d8`](https://github.com/dantp-ai/gym-puddle/commit/6ac75d8) (initial example) and [`140d1a8`](https://github.com/dantp-ai/gym-puddle/commit/140d1a8) (tuning + GIFs).

Measured on `seed=43`, ~9 minutes on a laptop CPU:

| policy        | mean_return | mean_length | success_rate |
|---------------|-------------|-------------|--------------|
| random        |    -3149.92 |      477.98 |       12.00% |
| discrete-sac  |    -7017.06 |      395.22 |       62.00% |

Demo GIFs at `examples/outputs/random.gif` and `examples/outputs/discrete_sac.gif`.

## Goal

Add a tutorial-grade example that trains and evaluates a modern off-policy actor-critic agent on `PuddleWorld-v0`, complementing `examples/0_uniform_policy.py`.

This is **not** a faithful reproduction of the 2012 Off-PAC algorithm (which uses linear features + tile coding + GTD(λ) — a fundamentally different algorithm). Instead it uses **Discrete SAC**, the canonical modern off-policy actor-critic for discrete action spaces.

## Algorithm and library

- **Algorithm:** Discrete SAC. PuddleWorld is `Discrete(5)`, which rules out DDPG/TD3 and continuous SAC.
- **Library:** [`torchRL`](https://github.com/pytorch/rl). Provides `DiscreteSACLoss`, `SoftUpdate`, `SyncDataCollector`, `TensorDictReplayBuffer` out of the box. Tianshou and vanilla PyTorch were considered; torchRL chosen by preference, and neither alternative implements 2012 Off-PAC either.

## Scope

A single self-contained script `examples/1_off_pac_discrete_sac.py` (~250 LOC), structured as small functions wired together in `main()`:

| Function | Returns | Notes |
|---|---|---|
| `build_env` | torchRL env | Wraps `PuddleWorld-v0`; one for train, one for eval. |
| `build_actor` | `ProbabilisticActor` | MLP `2 → 64 → 64 → 5`, `OneHotCategorical`. |
| `build_qvalue` | `TensorDictModule` | MLP `2 → 64 → 64 → 5`, outputs `action_value`. |
| `build_loss` | `(loss_module, target_updater)` | `DiscreteSACLoss` (2 Q-nets, `delay_qvalue=True`, `target_entropy_weight=0.3`) + `SoftUpdate`. |
| `train` | trained actor | Collector → buffer → N gradient steps → soft target update. |
| `evaluate` | `EvalMetrics` | Greedy / argmax rollout for `eval_episodes` episodes. |
| `random_baseline` | `EvalMetrics` | Same eval loop with `env.action_space.sample()`. |
| `record_gif` | `(return, length, terminated)` | Rolls out one episode on a raw gym env with `render_mode="rgb_array"`, writes a GIF via imageio. |
| `main` | — | Wires everything, prints comparison table, records GIFs, asserts perf bar. |

Hyperparameters as constants at the top of the file. The originally-spec'd values (50K frames, lr=3e-4, auto target entropy) didn't drive the agent to reach the goal — tuning landed at the values below:

```
total_frames           = 300_000      frames_per_batch       =    200
init_random_frames     =   1_000      batch_size             =    256
buffer_size            = 300_000      lr                     =   1e-4
gamma                  =    0.99      tau                    =  0.005
hidden_dim             =      64      reward_scale           =    0.1
target_entropy_weight  =     0.3      updates_per_batch      =    100
eval_episodes          =      50      max_episode_steps      =    500
seed                   =      43
```

## Acceptance criteria

**Functional**
- [x] `examples/1_off_pac_discrete_sac.py` runs end-to-end with `source .venv/bin/activate && python examples/1_off_pac_discrete_sac.py`.
- [x] Wall-clock ≤ 10 minutes on a typical laptop CPU. *(measured: ~9 minutes on Apple Silicon CPU)*
- [x] Prints a two-row comparison table over the same eval episodes.

**Performance bar**
- [ ] ~~`discrete_sac.mean_return > random.mean_return + 0.5 × |random.mean_return|` (≥ 50% improvement over random's magnitude).~~ *Superseded — with `target_entropy_weight=0.3` the agent commits to a direct route through the puddles, so `mean_return` worsens even as `success_rate` jumps dramatically. The success-rate bar below replaces this one.*
- [x] **Replacement bar:** `discrete_sac.success_rate ≥ 30%` AND `≥ 2× random.success_rate`. *(measured: 62% vs 12%, a 5.2× ratio)*
- [x] Asserted at the end of `main()` so a regression fails loudly.
- [x] Reported on a fixed seed (`seed=43`).

**Reproducibility**
- [x] RNGs seeded across env (`reset(seed)`, `action_space.seed(seed)`), `torch.manual_seed`, `numpy.random.seed`, and the torchRL collector seed.
- [x] Same seed → same printed metrics on CPU.

**Code quality**
- [x] Passes `ruff check examples/1_off_pac_discrete_sac.py`.
- [x] Passes `mypy examples/1_off_pac_discrete_sac.py` under repo config.
- [x] Module docstring explains algorithm choice and points at the Off-PAC paper.

**Packaging**
- [x] New optional-deps group in `pyproject.toml`:
  ```toml
  [project.optional-dependencies]
  rl-examples = [
      "torch>=2.2",
      "torchrl>=0.6",
      "tensordict>=0.6",
      "imageio>=2.34",
  ]
  ```
  *(`imageio` added to support GIF rendering of the trained agent.)*
- [x] `uv.lock` refreshed.
- [x] `README.md` gains a short "Off-policy actor-critic example" section with install + run commands, plus embedded GIFs.

**Tests**
- [x] Smoke test at `tests/examples/test_off_pac_example.py`:
  - Imports the module without running `main()`.
  - Calls `build_env`, `build_actor`, `build_qvalue`, `build_loss`; verifies shapes/specs.
  - `pytest.importorskip("torchrl")` so CI without the extra still passes.
- No full-training test (too slow, too seed-fragile).

## Out of band (delivered as part of this PR)

- **Env reset bug fix.** `PuddleEnv.reset()` computed the initial observation via `_get_initial_obs()` but never assigned it to `self.pos`, so subsequent `step()` calls operated on the previous episode's final position. Surfaced while building the example; fixed with a one-line change in `gym_puddle/envs/puddle_env.py` and guarded by a regression test in `tests/envs/test_env_implementation.py`.
- **Build-backend fix.** `pyproject.toml`'s `build-backend = "setuptools*"` is not a valid PEP 517 module name and blocked `uv sync` from a clean checkout; restored to `setuptools.build_meta`.
- **mypy override.** Added `[[tool.mypy.overrides]]` for `torchrl.*`, `tensordict.*`, and `numba` (none ship `py.typed`), so mypy passes when those deps aren't installed.
- **Demo GIFs.** `examples/outputs/random.gif` (~230 KB) and `examples/outputs/discrete_sac.gif` (~72 KB) generated by the example itself (no separate script), embedded in the README.

## Out of scope (candidates for follow-up PRs)

- Faithful 2012 Off-PAC reproduction (linear + tile coding + GTD(λ), ρ-weighted actor with eligibility traces).
- Logging backends (TensorBoard / Weights & Biases).
- Checkpointing + `--train` / `--eval` CLI modes.
- Learning-curve plots (matplotlib).
- Multi-seed sweep with mean ± standard error.
- Companion notebook walkthrough.
- GPU support / device flag.
- Hyperparameter sweep harness (Optuna).
- Additional algorithms (Discrete IQN, DQN, V-trace/IMPALA) as side-by-side comparators.
- Render the goal marker for `goal=[1.0, 1.0]` (the default goal coordinates put the red rect's top-left at the canvas corner so it's not visible in the GIF).
